Web Inspector: provide an option for controlling whether local overrides entirely replace or have some passthrough

dcrousso opened the pull request:
#### 716557a92a13730d3648c09b0bf8d18cc691a735
<pre>
Web Inspector: provide an option for controlling whether local overrides entirely replace or have some passthrough
<a href="https://bugs.webkit.org/show_bug.cgi?id=241990">https://bugs.webkit.org/show_bug.cgi?id=241990</a>

Reviewed by Patrick Angle.

* Source/WebInspectorUI/UserInterface/Models/LocalResourceOverride.js:
(WI.LocalResourceOverride.prototype.create):
(WI.LocalResourceOverride.fromJSON):
(WI.LocalResourceOverride.prototype.toJSON):
(WI.LocalResourceOverride.prototype.get isPassthrough): Added.
Add a `isPassthrough` member that indicates whether to include original request/response data (where
applicable/possible, e.g. headers).

* Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
(WI.NetworkManager.prototype.async requestIntercepted):
(WI.NetworkManager.prototype.async responseIntercepted):
Use `isPassthrough` to decide whether to return the original `request`/`response` data (or `undefined`
if the argument is optional) or the data from the `WI.LocalResourceOverride` (or `WI.LocalResource`).

* Source/WebInspectorUI/UserInterface/Views/LocalResourceOverridePopover.js:
(WI.LocalResourceOverridePopover):
(WI.LocalResourceOverridePopover.prototype.get serializedData):
(WI.LocalResourceOverridePopover.prototype.show):
* Source/WebInspectorUI/UserInterface/Views/LocalResourceOverridePopover.css:
(.popover .local-resource-override-popover-content &gt; table &gt; tr.options &gt; th): Added.
(.popover .local-resource-override-popover-content &gt; table &gt; tr.options &gt; td &gt; label:not([hidden])): Added.
(.popover .local-resource-override-popover-content .options td): Deleted.
Add UI at the bottom of the local override popover for controlling `isPassthrough`.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* LayoutTests/http/tests/inspector/network/intercept-request-properties.html:
* LayoutTests/http/tests/inspector/network/intercept-request-properties-expected.txt:
* LayoutTests/http/tests/inspector/network/local-resource-override-basic.html:
* LayoutTests/http/tests/inspector/network/local-resource-override-basic-expected.txt:

Canonical link: <a href="https://commits.webkit.org/251884@main">https://commits.webkit.org/251884@main</a>
</pre>

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/LocalResourceOverridePopover.js:
```diff
@@ -532,14 +543,19 @@ WI.LocalResourceOverridePopover = class LocalResourceOverridePopover extends WI.
+                isPassthroughLabelText.textContent = WI.UIString("Include original request data");
             }
 
             if (isResponse) {
```
`else if (isReponse)`?

dcrousso replied:
i'd personally like to keep this as-is in case we decide to add a way for a single local override to affect both request and response in the future, but i recognize that there will likely have to be other things that change then too (e.g. need more than one "Headers" table) so i'll change

patrickangle commented on LayoutTests/http/tests/inspector/network/intercept-request-properties.html:
```diff
@@ -164,12 +164,37 @@
     addTestCase({
-        name: "Network.interceptRequest.Headers.Passthrough",
+        name: "Network.interceptRequest.Headers.Empty",
         description: "Tests request headers passthrough",
```
Nit: This test case as well as the two new test cases have identical descriptions. It might be nice to be a little more specific about what each one is testing (e.g. how they differ).

dcrousso replied:
oh oops yes will change 😅

patrickangle commented on Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
```diff
@@ -996,10 +1007,19 @@ WI.NetworkManager = class NetworkManager extends WI.Object
-                    headers: localResource.responseHeaders,
+                    statusText: (function() {
+                        if (!isNaN(localResource.statusCode)) {
+                            if (localResource.statusText ?? false)
```
Shouldn't `localResource.statusText` itself be enough here? I think this could also be simplified down to `return localResource.statusText ?? WI.HTTPUtilities.statusTextForStatusCode(localResource.statusCode)`

dcrousso replied:
I was trying to match existing behavior, but I kinda agree in that there's not really a reason for us to be able to provide an arbitrary `statusText` even if there's no `statusCode`.

I think the logic was that the `statusCode` is kinda more important than the `statusText`, so if the former is not provided then we kinda dropped the latter on the floor.

patrickangle commented on Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
```diff
@@ -1032,10 +1055,30 @@ WI.NetworkManager = class NetworkManager extends WI.Object
+                            if (localResource.statusText ?? false)
+                                return localResource.statusText;
+
+                            return WI.HTTPUtilities.statusTextForStatusCode(localResource.statusCode);
```
Ditto :1014

dcrousso committed:
Web Inspector: provide an option for controlling whether local overrides entirely replace or have some passthrough
https://bugs.webkit.org/show_bug.cgi?id=241990

Reviewed by Patrick Angle.

* Source/WebInspectorUI/UserInterface/Models/LocalResourceOverride.js:
(WI.LocalResourceOverride.prototype.create):
(WI.LocalResourceOverride.fromJSON):
(WI.LocalResourceOverride.prototype.toJSON):
(WI.LocalResourceOverride.prototype.get isPassthrough): Added.
Add a `isPassthrough` member that indicates whether to include original request/response data (where
applicable/possible, e.g. headers).

* Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
(WI.NetworkManager.prototype.async requestIntercepted):
(WI.NetworkManager.prototype.async responseIntercepted):
Use `isPassthrough` to decide whether to return the original `request`/`response` data (or `undefined`
if the argument is optional) or the data from the `WI.LocalResourceOverride` (or `WI.LocalResource`).

* Source/WebInspectorUI/UserInterface/Views/LocalResourceOverridePopover.js:
(WI.LocalResourceOverridePopover):
(WI.LocalResourceOverridePopover.prototype.get serializedData):
(WI.LocalResourceOverridePopover.prototype.show):
* Source/WebInspectorUI/UserInterface/Views/LocalResourceOverridePopover.css:
(.popover .local-resource-override-popover-content > table > tr.options > th): Added.
(.popover .local-resource-override-popover-content > table > tr.options > td > label:not([hidden])): Added.
(.popover .local-resource-override-popover-content .options td): Deleted.
Add UI at the bottom of the local override popover for controlling `isPassthrough`.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* LayoutTests/http/tests/inspector/network/intercept-request-properties.html:
* LayoutTests/http/tests/inspector/network/intercept-request-properties-expected.txt:
* LayoutTests/http/tests/inspector/network/local-resource-override-basic.html:
* LayoutTests/http/tests/inspector/network/local-resource-override-basic-expected.txt:

Canonical link: https://commits.webkit.org/251884@main
